ssl/ssl_lib.c

```diff
@@ -2829,6 +2829,21 @@ void SSL_CTX_set_npn_select_cb(SSL_CTX *ctx,
 }
 #endif
 
+static int alpn_value_ok(const unsigned char *protos, unsigned int protos_len)
+{
+    const unsigned char *limit;
+
+    if (protos_len < 2 || protos == NULL)
+        return 0;
+
+    limit = protos + protos_len;
+    while (protos < limit) {
+        if (*protos == 0)
+            return 0;
+        protos += 1 + *protos;
+    }
+    return protos == limit;
```

Comment:
In the event the buffer is invalid then isn't there the potential for undefined behaviour? e.g. lets say the first byte in `protos` is 255, but `protos_len` is something less than that (say 3), then `protos` is now pointing more than one byte beyond the end of the allocated buffer which is undefined.

A different way to write this might be:
````
do {
    if (*protos == 0 || limit - protos < *protos + 1)
        return 0;
    protos += 1 + *protos;
} while (protos < limit);

return 1;
````

Reply:
No, the `protos < limit` ensures that case won't happen. For your example, assuming protos = 10000, and protos_len = 3, then limit = 10003. The first iteration, *protos = 255, then protos is incremented by that value + 1, so protos = 10000 + 1 + 255 = 10256, which fails the while check on the next iteration. Since protos is not dereferenced after the while loop, you won't access an undefined location.

Reply:
Note that the bad4 unit test checks for this, although no guarantee that it will be accessing an undefined location.

Reply:
> Since protos is not dereferenced after the while loop, you won't access an undefined location.

I'm not suggesting that the location is dereferenced (which would be a potential crash). I'm saying that just having the pointer pointing more than one byte beyond the end of the array is undefined in itself - regardless of whether it is dereferenced.

Reply:
For example:

````
int arr[4] = {0, 1, 2, 3};
int *p = arr + 5;  // undefined behavior for indexing out of bounds
````
From https://en.wikipedia.org/wiki/Undefined_behavior

Reply:
The example is indexing beyond the end of a locally defined array, vs. generic pointer arithmetic.
The compiler does not know the size of the object referenced by `protos`, and it is not able to semantically link `protos_len` to it. Any arithmetic on the pointers is valid; since `limit` is based on `protos`, the compiler treats them as the same object with unknown size. This is a much different case than when the compiler knows the size of the array (as in the example).
P.S. I'm arguing that this is a bad example, a citation of the relevant standards would be better, but we all know that those standards documents cost money... :(

Reply:
Changing it to not use pointers, but `protos_len` directly, instead.

Reply:
```
unsigned int idx;
unsigned char next;

for (idx = 0; idx < protos_len; idx += next) {
   next = protos[idx];
   if (next == 0)
       return 0;
}
return idx == protos_len;
```
No pointer arithmetics, no problem.

Reply:
Yes, similar idea.

Reply:
> Any arithmetic on the pointers is valid

My understanding of the standards is that this is not the case. Pointer arithmetic is defined in the standards in terms of pointing to an array object. It does not matter whether that array object is dynamically or statically allocated (e.g. see the answer here: https://stackoverflow.com/questions/54462378/undefined-behavior-with-pointer-arithmetic-on-dynamically-allocated-memory)

The standard says this about pointer arithmetic:
> If both the pointer operand and the result point to elements of the same array object, or one past the last element of the array object, the evaluation shall not produce an overflow; otherwise, the behavior is undefined.

I am fairly sure we've had ubsan issues related to this in the past.

Reply:
I've changed it to use an `idx` pointer comparing to `protos_len`, so the pointer issue is now moot.